### ISSUE TITLE
Make a colorbar for datashaded images?

### ISSUE THREAD

**Issue by timothydmorton** [ioam/holoviews#2102]
I know there might be another issue about this around somewhere, but I couldn't find it, so I thought I'd write a new one.  I understand that a colorbar for a datashaded image can't be handled the same way as other colorbars, since there's no way to get the data back from the image once it's created.  I was thinking that one possible solution might be to generate an inset *image* of a colorbar at the same time each datashaded view is generated, and superimpose it onto the image in some specified location.  Or even something as simple as optional (and discreet) labeling of the max/min of the scale being used at any time would be useful.  And it would be nice to have an option for this sort of thing within the `datashade` operation, rather than having to hack it up using `dynamic=False` and passing streams manually, etc.

**Comment by jbednar** [comment]
Yes, those are all things that we really need to do.  It's a bit awkward to put it together, and it's been on hold pending other urgent things, but I hope to get back to that soon.  To start with we just need to preserve the information about what transformations were done, which I made a plan for at the datashader site (https://github.com/bokeh/datashader/issues/126).  Once that's in place, HoloViews can read out that information and generate fully informative colorbars, hover information, and so on.  Hopefully this is something I can start on after Thanksgiving; I'm focusing on getting basic documentation (https://github.com/bokeh/datashader/issues/437) in place between now and then.

**Comment by timothydmorton** [comment]
Awesome, thanks for the update, and thanks again for all your great work on all of this!

**Comment by timothydmorton** [comment]
Question regarding a possible temporary workaround for this.  If I overlay a decimated plot (made invisible with `alpha=0`), that gives me hover information for a subset of the points.  I just realized this can also give me a colorbar.  However, I find that the the color scaling of datashader seems to be different from that of the underlying invisible `Points`.  If they used the same scaling, then this would be a great temporary workaround to the issue of a colorbar for a datashaded plot.  Is there any way to ensure they both use the same color scaling (modulo tiny variations that might happen because of the randomness of the decimating)?  Or should they be doing this already, and so I'm seeing a bug if they don't seem to match?

**Comment by timothydmorton** [comment]
Also, another suggestion regarding the hoped-for new datashader features--- perhaps y'all have thought of this already, but might it be worth it to have a datashaded plot be its own `Element` (i.e., distinct from the more general `RGB`)?  I don't know what additional complications that might bring, but maybe worth thinking about?

**Comment by jbednar** [comment]
Yes, you could ensure that they use the same scaling, by providing explicit scales to everything and using linear scaling.  E.g. the shade function accepts a span argument, to set the value ranges, and you can tell the dummy item the same range.  But this is definitely something we need to address properly sooner rather than later, and it's on my calendar to get to after Thanksgiving.

**Comment by jlstevens** [comment]
As much as we would love this, I doubt it will be working the way we want by 1.10 so I'm assigning this issue to the 2.0 milestone.

**Comment by rabernat** [comment]
I just hit this issue. As much as I love datashader, the lack of an easy way to get a colorbar is a major downside.

What workarounds do people use for this?

**Comment by jbednar** [comment]
From within holoviews, usually we just use `rasterize()` rather than `datashade()`, which lets Bokeh do the colormapping, so that there will be a colorbar.  The main downside is that Bokeh doesn't support histogram equalization.

**Comment by TheoMathurin** [comment]
If i'm not mistaken, another downside comes from the fact that Bokeh and Datashader do not handle log mapping the same way. Namely Datashader takes the data, substracts the minimum and then applies `np.log1p`. This allows for a much wider use of logarithmic scales compared to letting Bokeh handle the mapping, although you still have to build a custom colorbar that matches the colormapping that has been applied.

Also, you can't apply dynspread to rasterize.
